File: Find the K-th Character in String Game II/Find the K-th Character in String Game II.py
# Solution for Find the K-th Character in String Game II in PY

class Solution:
    def kthCharacter(self, k: int, operations: List[int]) -> str:
        if k == 1:
            return 'a'

        current_length = 1
        reduced_k = -1
        operation_type = 0

        for i in range(len(operations)):
            current_length *= 2
            if current_length >= k:
                operation_type = operations[i]
                reduced_k = k - current_length // 2
                break

        prev_char = self.kthCharacter(reduced_k, operations)

        if operation_type == 0:
            return prev_char
        else:
            return 'a' if prev_char == 'z' else chr(ord(prev_char) + 1)

File: Find the K-th Character in String Game II/Find the K-th Character in String Game II.java
// Solution for Find the K-th Character in String Game II in JAVA

class Solution {
    public char kthCharacter(long k, int[] operations) {
        if (k == 1) return 'a';

        long currentLength = 1;
        long reducedK = -1;
        int operationType = 0;

        for (int i = 0; i < operations.length; i++) {
            currentLength *= 2;
            if (currentLength >= k) {
                operationType = operations[i];
                reducedK = k - currentLength / 2;
                break;
            }
        }

        char prevChar = kthCharacter(reducedK, operations);

        if (operationType == 0) {
            return prevChar;
        } else {
            return (prevChar == 'z') ? 'a' : (char)(prevChar + 1);
        }
    }
}
